### ISSUE TITLE
Some confusion about SDWebImageDownloaderLIFOExecutionOrder

### ISSUE THREAD

**Issue by SmithJackyson** [SDWebImage/SDWebImage#2823]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 11 / Xcode 10
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

The execution order (last-in-first-out) of SDWebImageDownloaderOperation depends on property "lastAddedOperation" in SDWebImageDownloader, but it looks like something's wrong. The operation is removed from "URLOperations" when it's complete, and the "lastAddedOperation" will be nil. That means the new operation will execute like first-in-first-out. For example:(assume the downloadQueue.maxConcurrentOperationCount = 1)
there are 7 operation added into downloadQueue, the execution order will like:
7->6->5->4->3->2->1
the "lastAddedOperation" is operation 7, now the operation 7 is removed because it is complete, and this time new 7 operation are added into downloadQueue, the execution order will be:
6->5->4->3->2->1->14->13->12->11->10->9->8

I want to know if the code design is like this.

**Comment by dreampiggy** [comment]
What's the expected behavior for this LIFO order ? Actually this code was written in 2012, and does not changed (using a weak referenece) from 7 years ago. See:
https://github.com/SDWebImage/SDWebImage/blame/67f943b0faec54ccbe0ef446ea45f5ac0fbf17a1/SDWebImage/SDWebImageDownloader.m

Do you means `14 -> 13 -> 12 -> 11 -> 10 -> 9 -> 8 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1` ?

If so, we can design to change this behavior in 5.2.0.

**Comment by SmithJackyson** [comment]
> What's the expected behavior for this LIFO order ? Actually this code was written in 2012, and does not changed (using a weak referenece) from 7 years ago. See:
> https://github.com/SDWebImage/SDWebImage/blame/67f943b0faec54ccbe0ef446ea45f5ac0fbf17a1/SDWebImage/SDWebImageDownloader.m
> 
> Do you means `14 -> 13 -> 12 -> 11 -> 10 -> 9 -> 8 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1` ?
> 
> If so, we can design to change this behavior in 5.2.0.

yes, the execution order `14 -> 13 -> 12 -> 11 -> 10 -> 9 -> 8 -> 6 -> 5 -> 4 -> 3 -> 2 -> 1` will be right.

**Comment by dreampiggy** [comment]
Should be solved by #2852.
